### PR TITLE
[clang][test] add TestLanguage.def to specify all tested language versions

### PR DESCRIPTION
**Pull request by 5chmidti**
Adds a def file to have a single location where tested language versions are specified. Removes the need to update multiple locations in the testing infrastructure to add a new language version to be tested. Test instatiation can now include all languages without needing to specify them.
This patch also adds pretty printing for instantiated test names. That means, that a test instantiated with C++23 will have the name `...TestSuite/TestName/CXX23` instead ending with some number (index of the argument for instantiation of the test), which provides a better experience when encountering a test failure with a specific language version. The suffix will also contain an `_win` if the target contains `win`.

### CODE REVIEW

File: clang/include/clang/Testing/TestLanguage.def
Comment (AaronBallman) [inline clang/include/clang/Testing/TestLanguage.def:22]: It would be helpful to have a comment at the top of the file explaining the macro arguments

File: clang/include/clang/Testing/TestLanguage.def
Comment (AaronBallman) [inline clang/include/clang/Testing/TestLanguage.def:26]: Add c2y now that it was added to trunk?

Comment (5chmidti) [reply clang/include/clang/Testing/TestLanguage.def:33]: done

File: clang/include/clang/Testing/TestLanguage.def
Comment (5chmidti) [reply clang/include/clang/Testing/TestLanguage.def:29]: done

**Comment by 5chmidti** [comment]
rebased on trunk

**Comment by 5chmidti** [comment]
ping

**Comment by 5chmidti** [comment]
rebase + ping. @AaronBallman could you check if this is okay now? I could merge the AST matcher doc testing stack if that is the case

File: clang/include/clang/Testing/TestClangConfig.h
Comment (Sirraide) [reply clang/include/clang/Testing/TestClangConfig.h:123]: Ah, actually, no, this is passing in the version not the version index, so I’m not sure it would.

File: clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp
Comment (AaronBallman) [reply clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp:1227]: C99OrEarlier, but good catch!

**Review by AaronBallman — COMMENTED** [review]
modulo things brought up by @Sirraide, this is looking good!

File: clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp
Comment (5chmidti) [reply clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp:1227]: The idea was that previously only `Lang_C89` and `Lang_C99` were even declared in the `TestLanguage` enum, so I enabled the test for all `C` languages in this case.

File: clang/include/clang/Testing/TestClangConfig.h
Comment (5chmidti) [reply clang/include/clang/Testing/TestClangConfig.h]: But it could be expressed when `isCXX(int)` is introduced:

```c+v
  bool isCXXOrEarlier(int MaximumStdVersion) const {
    return isCXX() &&
           (isCXX(MaximumStdVersion) || !isCXXOrLater(MaximumStdVersion));
  }
```

I've pushed a change that does just that

File: clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp
Comment (AaronBallman) [reply clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp:1227]: Ah, okay, that's a reasonable change then, thank you!

**Review by AaronBallman — APPROVED** [review]
LGTM, unless @Sirraide has other concerns. Thank you for the improvement!

**Comment by 5chmidti** [comment]
@Sirraide because this is touching the unit test infra, I held of on merging so you could take another look once you were back.

**Review by Sirraide — APPROVED** [review]
Ah, no, I didn’t have anything else to comment on; lgtm too

**Pull request merged**
